src/Google.Api.Gax/Api/Gax/PathTemplate.cs

```diff
@@ -0,0 +1,439 @@
+        /// </para>
+        /// </remarks>
+        /// <param name="resourceIds">The resource IDs to use to populate the parameters in this template.</param>
+        /// <returns>The string representation of the resource name.</returns>
+        public string ExpandWithService(string serviceName, string[] resourceIds)
+        {
+            ValidateResourceIds(resourceIds);
+            ValidateServiceName(serviceName, nameof(serviceName));
+            return ReplaceParameters(serviceName, resourceIds);
+        }
+
+        /// <summary>
+        /// Returns a string representation of the template with parameters replaced by resource IDs.
+        /// </summary>
+        /// <param name="resourceIds">Resource IDs to interpolate the template with. Expected to have been validated already.</param>
+        internal string ReplaceParameters(string serviceName, string[] resourceIds)
+        {
+            int nextIdIndex = 0;
+            var result = new StringBuilder();
+            if (serviceName != null)
+            {
+                result.Append("//").Append(serviceName);
+            }
+            foreach (var segment in _segments)
+            {
+                if (result.Length != 0)
+                {
+                    result.Append('/');
+                }
+                switch (segment.Kind)
+                {
+                    case SegmentKind.Literal:
+                        result.Append(segment.Value);
+                        break;
+                    case SegmentKind.Wildcard:
+                        result.Append(resourceIds[nextIdIndex++]);
+                        break;
+                    case SegmentKind.PathWildcard:
+                        string path = resourceIds[nextIdIndex++];
+                        result.Append(path);
+                        if (path == "" && result.Length > 0)
+                        {
+                            // Swallow any slash that we've already added for this.
+                            result.Length--;
+                        }
+                        break;
+                }
+            }
+            return result.ToString();
+        }
+
+        /// <summary>
+        /// Attempts to parse the given resource name against this template, returning <c>null</c> on failure.
+        /// </summary>
+        /// <remarks>
+        /// Although this method returns <c>null</c> if a name is passed in which doesn't match the template,
+        /// it still throws <see cref="ArgumentNullException"/> if <paramref name="name"/> is null, as this would
+        /// usually indicate a programming error rather than a data error.
+        /// </remarks>
+        /// <param name="name">The resource name to parse against this template.</param>
+        /// <returns>The parsed name as a <see cref="ResourceName"/>, or <c>null</c> if the name does not match this
+        /// template.</returns>
+        public ResourceName TryParseName(string name)
```

Comment:
That is not the usual pattern in the BCL, it would typically be:
```C#
bool TryParseName(string name, out ResourceName result);
```
I don't particularly like that pattern either but is something to consider with respect of being idiomatic.